File: lib/AbstractLayer/AbsLayerHDWallet.swift
//
//  AbsLayerHDWallet.swift
//  lib
//
//  Created by mengxk on 2019/3/23.
//  Copyright © 2019 Elastos. All rights reserved.
//

extension AbstractLayer {
  final class HDWallet {
    private init() {}
    
    public static func GetCoinType(obj: Int64) -> Int {
      let coinType = AbsLayer_HDWallet_GetCoinType(obj)
      
      return Int(coinType)
    }
    
    public static func SendTransaction(obj: Int64, transactions: [Int64],
                                       memo: String, seed: String, chain: String) -> String? {
      let transactionsPtr = Array.ToUnsafeMutablePointer(data: transactions)
      let memoPtr = String.ToUnsafeMutablePointer(data: memo)
      let seedPtr = String.ToUnsafeMutablePointer(data: seed)
      let chainPtr = String.ToUnsafeMutablePointer(data: chain)
      
      let txidPtr = AbsLayer_HDWallet_SendTransaction(obj, transactionsPtr, Int32(transactions.count),
                                                      memoPtr, seedPtr, chainPtr)
      let txid = String.FromUnsafeMutablePointer(data: txidPtr)
      AbstractLayer_FreeBuf(txidPtr)
      
      return txid
    }

    public static func GetAddress(obj: Int64, chain: Int, index: Int) -> String? {
      let addrPtr = AbsLayer_HDWallet_GetAddress(obj, Int32(chain), Int32(index))
      let addr = String.FromUnsafeMutablePointer(data: addrPtr)
      AbstractLayer_FreeBuf(addrPtr)
      
      return addr
    }
    
    public static func GetPublickey(obj: Int64, chain: Int, index: Int) -> String? {
      let pubKeyPtr = AbsLayer_HDWallet_GetPublickey(obj, Int32(chain), Int32(index))
      let pubKey = String.FromUnsafeMutablePointer(data: pubKeyPtr)
      AbstractLayer_FreeBuf(pubKeyPtr)
      
      return pubKey
    }
    
    public static func GetBalance(obj: Int64, address: String) -> Int64 {
      let addressPtr = String.ToUnsafeMutablePointer(data: address)
      
      let balance = AbsLayer_HDWallet_GetBalance(obj, addressPtr)
      
      return balance
    }
    
    public static func GetBalanceEx(obj: Int64) -> Int64 {
      let balance = AbsLayer_HDWallet_GetBalanceEx(obj)
      
      return balance
    }
    
    public static func SyncHistory(obj: Int64) -> Int {
      let ret = AbsLayer_HDWallet_SyncHistory(obj)
      
      return Int(ret)
    }
    
    public static func GetHistoryCount(obj: Int64, address: String) -> Int {
      let addressPtr = String.ToUnsafeMutablePointer(data: address)
      
      let count = AbsLayer_HDWallet_GetHistoryCount(obj, addressPtr)
      
      return Int(count)
    }
    
    public static func GetHistory(obj: Int64, address: String, pageSize: Int, page: Int, ascending: Bool) -> String? {
      let addressPtr = String.ToUnsafeMutablePointer(data: address)
      
      let historyPtr = AbsLayer_HDWallet_GetHistory(obj, addressPtr, Int32(pageSize), Int32(page), ascending)
      let history = String.FromUnsafeMutablePointer(data: historyPtr)
      AbstractLayer_FreeBuf(historyPtr)
      
      return history
    }

    public static func GetUsedAddresses(obj: Int64) -> [String] {
      var usedAddresse = [String]()

      let usedAddressePtr = AbsLayer_HDWallet_GetUsedAddresses(obj)
      var idx = 0
      while(usedAddressePtr![idx] != nil) {
        let addr = String.FromUnsafeMutablePointer(data: usedAddressePtr![idx])
        usedAddresse.append(addr!)
        idx += 1
      }
      AbsLayer_HDWallet_FreeAddresses(usedAddressePtr)
      
      return usedAddresse
    }
    
    public static func GetUnUsedAddresses(obj: Int64, count: Int) -> [String] {
      var unusedAddresse = [String]()
      
      let unusedAddressePtr = AbsLayer_HDWallet_GetUnUsedAddresses(obj, Int32(count))
      var idx = 0
      while(unusedAddressePtr![idx] != nil) {
        let addr = String.FromUnsafeMutablePointer(data: unusedAddressePtr![idx])
        unusedAddresse.append(addr!)
        idx += 1
      }
      AbsLayer_HDWallet_FreeAddresses(unusedAddressePtr)
      
      return unusedAddresse
    }
    
    public static func Recover(obj: Int64) -> Int {
      let ret = AbsLayer_HDWallet_Recover(obj)
      return Int(ret)
    }

    public static func DestroyHDWallet(obj: Int64) {
      AbsLayer_HDWallet_DestroyHDWallet(obj)
    }
  }
}

File: lib/AbstractLayer/AbsLayerHDWallet.cpp
//
//  AbsLayerHDWallet.cpp
//  lib
//
//  Created by mengxk on 2019/3/23.
//  Copyright © 2019 Elastos. All rights reserved.
//

#include "AbsLayerHDWallet.hpp"

#include "HDWallet.h"
#include "Transaction.h"

int AbsLayer_HDWallet_GetCoinType(int64_t obj)
{
  if (obj == 0) {
    return 0;
  }
  std::shared_ptr<elastos::HDWallet>* wallet = reinterpret_cast<std::shared_ptr<elastos::HDWallet>*>(obj);
 
  return (*wallet)->GetCoinType();
}

char* AbsLayer_HDWallet_SendTransaction(int64_t obj, int64_t transactions[], int transactionsLen,
                                              const char* memo,  const char* seed, const char* chain)
{
  if (obj == 0) {
    return 0;
  }
  std::shared_ptr<elastos::HDWallet>* wallet = reinterpret_cast<std::shared_ptr<elastos::HDWallet>*>(obj);
  
  std::vector<elastos::Transaction> txs;
  for (int i = 0; i < transactionsLen; i++) {
    elastos::Transaction* tx = (elastos::Transaction*)transactions[i];
    if (tx == nullptr) {
       continue;
    }
    txs.push_back(*tx);
  }
  std::string txid;
  (*wallet)->SendTransaction(txs, memo, seed, txid, chain);

  char* txidBuf = strdup(txid.c_str());
  return txidBuf;
}

char* AbsLayer_HDWallet_GetAddress(int64_t obj, int chain, int index)
{
  if (obj == 0) {
    return 0;
  }
  std::shared_ptr<elastos::HDWallet>* wallet = reinterpret_cast<std::shared_ptr<elastos::HDWallet>*>(obj);
  
  std::string address = (*wallet)->GetAddress(chain, index);

  char* addressBuf = strdup(address.c_str());
  return addressBuf;
}

char* AbsLayer_HDWallet_GetPublickey(int64_t obj, int chain, int index)
{
  if (obj == 0) {
    return 0;
  }
  std::shared_ptr<elastos::HDWallet>* wallet = reinterpret_cast<std::shared_ptr<elastos::HDWallet>*>(obj);

  std::string pubKey = (*wallet)->GetPublicKey(chain, index);

  char* pubKeyBuf = strdup(pubKey.c_str());
  return pubKeyBuf;
}

int64_t AbsLayer_HDWallet_GetBalance(int64_t obj, const char* address)
{
  if (obj == 0) {
    return 0;
  }
  std::shared_ptr<elastos::HDWallet>* wallet = reinterpret_cast<std::shared_ptr<elastos::HDWallet>*>(obj);

  int64_t balance = (*wallet)->GetBalance(address);

  return balance;
}

int64_t AbsLayer_HDWallet_GetBalanceEx(int64_t obj)
{
  if (obj == 0) {
    return 0;
  }
  std::shared_ptr<elastos::HDWallet>* wallet = reinterpret_cast<std::shared_ptr<elastos::HDWallet>*>(obj);
  
  int64_t balance = (*wallet)->GetBalance();
  
  return balance;
}

int AbsLayer_HDWallet_SyncHistory(int64_t obj)
{
  if (obj == 0) {
    return 0;
  }
  std::shared_ptr<elastos::HDWallet>* wallet = reinterpret_cast<std::shared_ptr<elastos::HDWallet>*>(obj);
  
  int ret = (*wallet)->SyncHistory();
  
  return ret;
}

int AbsLayer_HDWallet_GetHistoryCount(int64_t obj, const char* address)
{
  if (obj == 0) {
    return 0;
  }
  std::shared_ptr<elastos::HDWallet>* wallet = reinterpret_cast<std::shared_ptr<elastos::HDWallet>*>(obj);
  
  int count = (*wallet)->GetHistoryCount(address);
  
  return count;
}

char* AbsLayer_HDWallet_GetHistory(int64_t obj, const char* address, int pageSize, int page, bool ascending)
{
  if (obj == 0) {
    return 0;
  }
  std::shared_ptr<elastos::HDWallet>* wallet = reinterpret_cast<std::shared_ptr<elastos::HDWallet>*>(obj);
  
  std::string histories;
 (*wallet)->GetHistory(address, pageSize, page, ascending, histories);
  
  char* historiesBuf = strdup(histories.c_str());
  return historiesBuf;
}

char** AbsLayer_HDWallet_GetUsedAddresses(int64_t obj)
{
  if (obj == 0) {
    return 0;
  }
  std::shared_ptr<elastos::HDWallet>* wallet = reinterpret_cast<std::shared_ptr<elastos::HDWallet>*>(obj);

  std::vector<std::string> addrs = (*wallet)->GetUsedAddresses();
  
  char** array = new char*[addrs.size() + 1];
  for(auto idx = 0; idx < addrs.size(); idx++) {
    array[idx] = strdup(addrs[idx].c_str());
  }
  array[addrs.size()] = nullptr;
  
  return array;
}

char** AbsLayer_HDWallet_GetUnUsedAddresses(int64_t obj, int count)
{
  if (obj == 0) {
    return 0;
  }
  std::shared_ptr<elastos::HDWallet>* wallet = reinterpret_cast<std::shared_ptr<elastos::HDWallet>*>(obj);
  
  std::vector<std::string> addrs = (*wallet)->GetUnUsedAddresses(count);
  
  char** array = new char*[addrs.size() + 1];
  for(auto idx = 0; idx < addrs.size(); idx++) {
    array[idx] = strdup(addrs[idx].c_str());
  }
  array[addrs.size()] = nullptr;

  
  return array;
}

void AbsLayer_HDWallet_FreeAddresses(char** addresses) {
  if(addresses == nullptr) {
    return;
  }
  
  int idx = 0;
  while(addresses[idx] != nullptr) {
    free(addresses[idx]);
    idx++;
  }
  
  delete[] addresses;
}

int AbsLayer_HDWallet_Recover(int64_t obj)
{
  if (obj == 0) {
    return 0;
  }
  std::shared_ptr<elastos::HDWallet>* wallet = reinterpret_cast<std::shared_ptr<elastos::HDWallet>*>(obj);
  
  int recover = (*wallet)->Recover();
  
  return recover;
}

void AbsLayer_HDWallet_DestroyHDWallet(int64_t obj)
{
  if (obj == 0) {
    return;
  }
  
  std::shared_ptr<elastos::HDWallet>* wallet = reinterpret_cast<std::shared_ptr<elastos::HDWallet>*>(obj);
  wallet->reset();
  delete wallet;
}
